### PR TITLE
feat(perf): use Mmap in `ManyCar::read_only_files`

### PR DESCRIPTION
**Pull request by hanabi1224**
## Summary of changes

Related to https://github.com/ChainSafe/forest/issues/3361

Changes introduced in this pull request:

- use Mmap in `ManyCar::read_only_files` for better performance.

P.S. benchmark shows the performance of using vanilla Mmap API is 2-5x better than using a Mmap wrapper that implements `positioned-io` traits, we could benchmark and discuss about it later

### `forest-tool benchmark graph-traversal` result

calibnet snapshot(local laptop):
```console
main branch:
traversed 2.84 GiB at 104.91 MiB/s in 00:00:27
traversed 2.84 GiB at 104.65 MiB/s in 00:00:27
traversed 2.84 GiB at 103.04 MiB/s in 00:00:28

feature branch:
traversed 2.84 GiB at 139.04 MiB/s in 00:00:20
traversed 2.84 GiB at 157.26 MiB/s in 00:00:18
traversed 2.84 GiB at 161.50 MiB/s in 00:00:17
```

mainnet snapshot(DO droplet):
```console
# first run
➜  car_db /root/forest-tool-main benchmark graph-traversal 3134862.forest.car.zst
  traversed 93.80 GiB at 25.01 MiB/s in 01:04:01
➜  car_db /root/forest-tool-mmap benchmark graph-traversal 3134862.forest.car.zst
  traversed 93.80 GiB at 95.05 MiB/s in 00:16:50
# second run
➜  car_db /root/forest-tool-main benchmark graph-traversal 3134862.forest.car.zst
  traversed 93.80 GiB at 36.10 MiB/s in 00:44:20
➜  car_db /root/forest-tool-mmap benchmark graph-traversal 3134862.forest.car.zst
  traversed 93.80 GiB at 93.92 MiB/s in 00:17:02
```

### `forest-tool snapshot validate [snapshot].forest.car.zst` result (with hyperfine)

calibnet(local laptop):
```
# hyperfine -w 1 "./target/release/forest-tool-mmap snapshot validate ~/fr/calibnet/paritydb/car_db/1692774391.forest.car.zst" "./target/release/forest-tool-main snapshot validate ~/fr/calibnet/paritydb/car_db/1692774391.forest.car.zst"

Benchmark 1: ./target/release/forest-tool-mmap snapshot validate ~/fr/calibnet/paritydb/car_db/1692774391.forest.car.zst
  Time (mean ± σ):     14.822 s ±  2.366 s    [User: 16.257 s, System: 0.449 s]
  Range (min … max):    9.717 s … 16.555 s    10 runs

Benchmark 2: ./target/release/forest-tool-main snapshot validate ~/fr/calibnet/paritydb/car_db/1692774391.forest.car.zst
  Time (mean ± σ):     16.309 s ±  2.409 s    [User: 16.436 s, System: 1.565 s]
  Range (min … max):    9.617 s … 18.128 s    10 runs

Summary
  './target/release/forest-tool-mmap snapshot validate ~/fr/calibnet/paritydb/car_db/1692774391.forest.car.zst' ran
    1.10 ± 0.24 times faster than './target/release/forest-tool-main snapshot validate ~/fr/calibnet/paritydb/car_db/1692774391.forest.car.zst'
```

mainnet(DO droplet):
```
# hyperfine -w 0 -r 3 "/root/forest-tool-mmap snapshot validate 3134862.forest.car.zst" "/root/forest-tool-main snapshot validate 3134862.forest.car.zst"

Benchmark 1: /root/forest-tool-mmap snapshot validate 3134862.forest.car.zst
  Time (mean ± σ):     1773.404 s ± 13.463 s    [User: 2020.588 s, System: 59.360 s]
  Range (min … max):   1765.198 s … 1788.942 s    3 runs

Benchmark 2: /root/forest-tool-main snapshot validate 3134862.forest.car.zst
  Time (mean ± σ):     2767.611 s ± 286.282 s    [User: 2123.710 s, System: 365.960 s]
  Range (min … max):   2444.817 s … 2990.734 s    3 runs

Summary
  /root/forest-tool-mmap snapshot validate 3134862.forest.car.zst ran
    1.56 ± 0.16 times faster than /root/forest-tool-main snapshot validate 3134862.forest.car.zst
```

## Reference issue to close (if applicable)

Closes

## Change checklist

- [x] I have performed a self-review of my own code,
- [x] I have made corresponding changes to the documentation,
- [x] I have added tests that prove my fix is effective or that my feature works (if possible),
- [x] I have made sure the [CHANGELOG][1] is up-to-date. All user-facing changes should be reflected in this document.

[1]: https://github.com/ChainSafe/forest/blob/main/CHANGELOG.md

### CODE REVIEW

**Comment by lemmih** [comment]
Let's use an environment variable to control whether to use file-io or mmap. And let's benchmark tipset validation. You can use hyperfine for this.

**Comment by hanabi1224** [comment]
> Let's use an environment variable to control whether to use file-io or mmap. And let's benchmark tipset validation. You can use hyperfine for this.

@lemmih Just to confirm, do you mean `forest-tool snapshot validate [foo].forest.car.zst` with default CLI options?

**Comment by lemmih** [comment]
> > Let's use an environment variable to control whether to use file-io or mmap. And let's benchmark tipset validation. You can use hyperfine for this.
> 
> @lemmih Just to confirm, do you mean `forest-tool snapshot validate [foo].forest.car.zst` with default CLI options?

Sure. You can adjust the parameters to whatever is most convenient for you.

**Review by sudo-shashank — APPROVED** [review]

**Comment by hanabi1224** [comment]
@lemmih `forest-tool snapshot validate` benchmark result updated in the PR description. Btw, which approach should be made default when the environment variable is not set?  Should I use mmap as default with `FOREST_CAR_LOADER_FILE_IO=1` as the switch or use file-io as default with `FOREST_CAR_LOADER_MMAP=1` as the switch?

File: src/utils/io/mmap.rs [hunk 1]
```diff
@@ -0,0 +1,37 @@
+// Copyright 2019-2023 ChainSafe Systems
+// SPDX-License-Identifier: Apache-2.0, MIT
+
+use std::io;
+
+use memmap2::MmapAsRawDesc;
+use positioned_io::{ReadAt, Size};
+
+/// Wrapper type of [`memmap2::Mmap`] that implements [`ReadAt`] and [`Size`]
+pub struct Mmap(memmap2::Mmap);
```
Comment (LesnyRumcajs) [inline src/utils/io/mmap.rs:10]: Let's have some tests for this, there tend to be dragons in buffers and indexing.

**Review by LesnyRumcajs — CHANGES_REQUESTED** [review]

File: src/utils/io/mmap.rs (hunk 1 above)
Comment (hanabi1224) [reply src/utils/io/mmap.rs:10]: Done

**Review by LesnyRumcajs — APPROVED** [review]

**Comment by hanabi1224** [comment]
@lemmih @LesnyRumcajs Environment variable and doc added

**Review by lemmih — APPROVED** [review]

File: documentation/src/environment_variables.md [hunk 2]
```diff
@@ -0,0 +1,10 @@
+# Environment variables
```
Comment (LesnyRumcajs) [inline documentation/src/environment_variables.md:1]: Good stuff

**Pull request merged**